File: src/libqstorageinfo.cpp
#include <QByteArray>
#include <QDir>
#include <QList>
#include <QStorageInfo>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qstorageinfo.h>
#include "libqstorageinfo.h"
#include "libqstorageinfo.hxx"

QStorageInfo* QStorageInfo_new() {
    return new QStorageInfo();
}

QStorageInfo* QStorageInfo_new2(libqt_string path) {
    QString path_QString = QString::fromUtf8(path.data, path.len);
    return new QStorageInfo(path_QString);
}

QStorageInfo* QStorageInfo_new3(QDir* dir) {
    return new QStorageInfo(*dir);
}

QStorageInfo* QStorageInfo_new4(QStorageInfo* other) {
    return new QStorageInfo(*other);
}

void QStorageInfo_OperatorAssign(QStorageInfo* self, QStorageInfo* other) {
    self->operator=(*other);
}

void QStorageInfo_Swap(QStorageInfo* self, QStorageInfo* other) {
    self->swap(*other);
}

void QStorageInfo_SetPath(QStorageInfo* self, libqt_string path) {
    QString path_QString = QString::fromUtf8(path.data, path.len);
    self->setPath(path_QString);
}

libqt_string QStorageInfo_RootPath(const QStorageInfo* self) {
    QString _ret = self->rootPath();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QStorageInfo_Device(const QStorageInfo* self) {
    QByteArray _qb = self->device();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _qb.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QStorageInfo_Subvolume(const QStorageInfo* self) {
    QByteArray _qb = self->subvolume();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _qb.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QStorageInfo_FileSystemType(const QStorageInfo* self) {
    QByteArray _qb = self->fileSystemType();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _qb.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QStorageInfo_Name(const QStorageInfo* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QStorageInfo_DisplayName(const QStorageInfo* self) {
    QString _ret = self->displayName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

long long QStorageInfo_BytesTotal(const QStorageInfo* self) {
    return static_cast<long long>(self->bytesTotal());
}

long long QStorageInfo_BytesFree(const QStorageInfo* self) {
    return static_cast<long long>(self->bytesFree());
}

long long QStorageInfo_BytesAvailable(const QStorageInfo* self) {
    return static_cast<long long>(self->bytesAvailable());
}

int QStorageInfo_BlockSize(const QStorageInfo* self) {
    return self->blockSize();
}

bool QStorageInfo_IsRoot(const QStorageInfo* self) {
    return self->isRoot();
}

bool QStorageInfo_IsReadOnly(const QStorageInfo* self) {
    return self->isReadOnly();
}

bool QStorageInfo_IsReady(const QStorageInfo* self) {
    return self->isReady();
}

bool QStorageInfo_IsValid(const QStorageInfo* self) {
    return self->isValid();
}

void QStorageInfo_Refresh(QStorageInfo* self) {
    self->refresh();
}

libqt_list /* of QStorageInfo* */ QStorageInfo_MountedVolumes() {
    QList<QStorageInfo> _ret = QStorageInfo::mountedVolumes();
    // Convert QList<> from C++ memory to manually-managed C memory
    QStorageInfo** _arr = static_cast<QStorageInfo**>(malloc(sizeof(QStorageInfo*) * _ret.length()));
    for (size_t i = 0; i < _ret.length(); ++i) {
        _arr[i] = new QStorageInfo(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.length();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

QStorageInfo* QStorageInfo_Root() {
    return new QStorageInfo(QStorageInfo::root());
}

void QStorageInfo_Delete(QStorageInfo* self) {
    delete self;
}

File: src/libqstorageinfo.zig
const C = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qstorageinfo.html
pub const qstorageinfo = struct {
    /// New constructs a new QStorageInfo object.
    ///
    ///
    pub fn New() ?*C.QStorageInfo {
        return C.QStorageInfo_new();
    }

    /// New2 constructs a new QStorageInfo object.
    ///
    /// ``` path: []const u8 ```
    pub fn New2(path: []const u8) ?*C.QStorageInfo {
        const path_str = C.struct_libqt_string{
            .len = path.len,
            .data = @constCast(path.ptr),
        };

        return C.QStorageInfo_new2(path_str);
    }

    /// New3 constructs a new QStorageInfo object.
    ///
    /// ``` dir: ?*C.QDir ```
    pub fn New3(dir: ?*anyopaque) ?*C.QStorageInfo {
        return C.QStorageInfo_new3(@ptrCast(dir));
    }

    /// New4 constructs a new QStorageInfo object.
    ///
    /// ``` other: ?*C.QStorageInfo ```
    pub fn New4(other: ?*anyopaque) ?*C.QStorageInfo {
        return C.QStorageInfo_new4(@ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#operator=)
    ///
    /// ``` self: ?*C.QStorageInfo, other: ?*C.QStorageInfo ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QStorageInfo_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#swap)
    ///
    /// ``` self: ?*C.QStorageInfo, other: ?*C.QStorageInfo ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QStorageInfo_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#setPath)
    ///
    /// ``` self: ?*C.QStorageInfo, path: []const u8 ```
    pub fn SetPath(self: ?*anyopaque, path: []const u8) void {
        const path_str = C.struct_libqt_string{
            .len = path.len,
            .data = @constCast(path.ptr),
        };
        C.QStorageInfo_SetPath(@ptrCast(self), path_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#rootPath)
    ///
    /// ``` self: ?*C.QStorageInfo, allocator: std.mem.Allocator ```
    pub fn RootPath(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QStorageInfo_RootPath(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#device)
    ///
    /// ``` self: ?*C.QStorageInfo, allocator: std.mem.Allocator ```
    pub fn Device(self: ?*anyopaque, allocator: std.mem.Allocator) []u8 {
        const _bytearray: C.struct_libqt_string = C.QStorageInfo_Device(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_bytearray));
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("Memory allocation failed");
        for (0.._bytearray.len) |_i| {
            _ret[_i] = _bytearray.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#subvolume)
    ///
    /// ``` self: ?*C.QStorageInfo, allocator: std.mem.Allocator ```
    pub fn Subvolume(self: ?*anyopaque, allocator: std.mem.Allocator) []u8 {
        const _bytearray: C.struct_libqt_string = C.QStorageInfo_Subvolume(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_bytearray));
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("Memory allocation failed");
        for (0.._bytearray.len) |_i| {
            _ret[_i] = _bytearray.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#fileSystemType)
    ///
    /// ``` self: ?*C.QStorageInfo, allocator: std.mem.Allocator ```
    pub fn FileSystemType(self: ?*anyopaque, allocator: std.mem.Allocator) []u8 {
        const _bytearray: C.struct_libqt_string = C.QStorageInfo_FileSystemType(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_bytearray));
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("Memory allocation failed");
        for (0.._bytearray.len) |_i| {
            _ret[_i] = _bytearray.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#name)
    ///
    /// ``` self: ?*C.QStorageInfo, allocator: std.mem.Allocator ```
    pub fn Name(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QStorageInfo_Name(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#displayName)
    ///
    /// ``` self: ?*C.QStorageInfo, allocator: std.mem.Allocator ```
    pub fn DisplayName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QStorageInfo_DisplayName(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#bytesTotal)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn BytesTotal(self: ?*anyopaque) i64 {
        return C.QStorageInfo_BytesTotal(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#bytesFree)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn BytesFree(self: ?*anyopaque) i64 {
        return C.QStorageInfo_BytesFree(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#bytesAvailable)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn BytesAvailable(self: ?*anyopaque) i64 {
        return C.QStorageInfo_BytesAvailable(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#blockSize)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn BlockSize(self: ?*anyopaque) i32 {
        return C.QStorageInfo_BlockSize(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#isRoot)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn IsRoot(self: ?*anyopaque) bool {
        return C.QStorageInfo_IsRoot(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#isReadOnly)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn IsReadOnly(self: ?*anyopaque) bool {
        return C.QStorageInfo_IsReadOnly(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#isReady)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn IsReady(self: ?*anyopaque) bool {
        return C.QStorageInfo_IsReady(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#isValid)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return C.QStorageInfo_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#refresh)
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn Refresh(self: ?*anyopaque) void {
        C.QStorageInfo_Refresh(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#mountedVolumes)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn MountedVolumes(allocator: std.mem.Allocator) []?*C.QStorageInfo {
        const _arr: C.struct_libqt_list = C.QStorageInfo_MountedVolumes();
        defer C.libqt_free(_arr.data);
        const _ret = allocator.alloc(?*C.QStorageInfo, _arr.len) catch @panic("Memory allocation failed");
        const _data: [*]?*C.QStorageInfo = @ptrCast(@alignCast(_arr.data));
        for (0.._arr.len) |_i| {
            _ret[_i] = _data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstorageinfo.html#root)
    ///
    ///
    pub fn Root() ?*C.QStorageInfo {
        return C.QStorageInfo_Root();
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QStorageInfo ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QStorageInfo_Delete(@ptrCast(self));
    }
};
